_source/_posts/2022-10-11-angular-dependency-injection.md

```diff
@@ -0,0 +1,332 @@
+
+ While having this level of configurability sounds unnecessarily complicated, it allows you to fine-tune which dependency to use in your consuming code. You can configure the `providers` array in other modules and Angular building blocks, such as components and directives. You'll get the provider you configured closest to the consuming code when you do so.
+
+Now that we have a quick overview of how to provide dependencies, let's review an integral piece of Angular's DI system, injection tokens.
+
+## Injection tokens in Angular
+
+Injection tokens allow us to have values and objects as dependencies. This means we can depend on a string, such as "Hello world!", and objects, which include configuration objects and global variables such as Web APIs. But injection tokens are even more remarkable because we can also create dependencies to constructs that don't have a runtime definition, such as interfaces! Let's take a look at an example using an injection token.
+
+Let's say we want to have a dependency on the bread emoji. You can create the bread emoji, register the token to Angular's DI system, and set up a default value like this:
+
+```ts
+const export BREAD_TOKEN = new InjectionToken<string>('bread', {
+  providedIn: 'root',
+  factory: () => '🍞'
+});
+```
+
+When you want to use the `BREAD_TOKEN`, you'd use the `@Inject` decorator:
+
+```ts
+@Component({
+  // standard component metadata here
+})
+export class MyBreadComponent {
+  constructor(@Inject(BREAD_TOKEN) private bread: string) { }
+}
+```
+
+Now we can access the 🍞 emoji from within the component! This might not seem like a big deal, but having injection tokens as a means to represent values and interfaces as dependencies is enormous! And it sets us up to leverage the power of Angular's DI system.
+
+We can use injection tokens along with configuring providers within Angular's DI system for more power and fine-grained control.
+
+## Configuring providers in Angular's Dependency Injection system
+
+You can configure the `providers` array to add fine-grained control to your providers. When combined with injection tokens, we can unleash a lot of power. But first, it's essential to know when it makes sense to do so. Always prefer the most straightforward, default way of registering a dependency and then use fine-grained control as needed.
+
+To configure the `providers` array, you add an object containing the instructions like this:
+
+```ts
+@NgModule({
+  imports: // stuff here
+  declarations: // stuff here
+  providers: [
+    { provide: MyService, howToProvide: OtherDependency }
+  ]
+})
+export class AppModule {
+}
+```
+
+The "how to provide" gives Angular-specific instructions on this dependency configuration. Then you can provide the other new dependency. Angular supports the following options for "how to provide":
```

Comment:
Any reason here not to use code markup for "how to provide" ?

Reply:
i was keeping the code markup for real options so that it wouldn't be confusing, but if you think it's clearer to use code markup here I can